app/models/message_period.rb

```diff
@@ -0,0 +1,183 @@
+  ResultLimit = 5000
+
+  # チャンネル識別子
+  #
+  # パラメータ名の都合で名前がchannelsでも識別子を表すことに注意。
+  # @return [String]
+  attr_accessor :channels
+  # 開始日
+  # @return [Time, nil]
+  #
+  # 名称はGoogle検索に準拠している。
+  #
+  # セッターでは、Time 型に変換できないときは nil になる。
+  attr_reader :since
+  # 終了日
+  # @return [Time, nil]
+  #
+  # 名称はGoogle検索に準拠している。
+  #
+  # セッターでは、Time 型に変換できないときは nil になる。
+  attr_reader :until
+
+  validate :until_must_not_be_less_than_since_if_both_exist
+  validate :until_set_now_datetime_if_not_exist
+
+  def initialize(*)
+    @channels = []
+    super
+  end
+
+  # 開始日時を設定する
+  #
+  # Time 型に変換できないときは nil になる。
+  # @param [#to_time] value 開始日時
+  def since=(value)
+    begin
+      @since = value.to_time
+    rescue
+      @since = nil
+    end
+  end
+
+  # 終了日を設定する
+  #
+  # Time 型に変換できないときは nil になる。
+  # @param [#to_time] value 終了日時
+  def until=(value)
+    begin
+      @until = value.to_time
+    rescue
+      @until = nil
+    end
+  end
+
+  # 属性のハッシュを返す
+  # @return [Hash]
+  def attributes
+    {
+      'channels' => @channels,
+      'since' => @since,
+      'until' => @until
+    }
+  end
+
+  # 指定したハッシュを使って属性を設定する
+  # @param [Hash] hash 属性の設定に使うハッシュ
+  # @return [void]
+  def attributes=(hash)
+    self.channels = hash['channels']
+    self.since = hash['since']
+    self.until = hash['until']
+  end
+
+  # 結果ページ向けの属性のハッシュを返す
+  # @return [Hash]
+  def attributes_for_result_page
+    {
+      'channels' => @channels.join(' '),
+      'since' => @since&.strftime('%F %T'),
+      'until' => @until&.strftime('%F %T')
+    }
+  end
+
+  # 結果ページのパラメータのハッシュを使って属性を設定する
+  # @param [Hash] hash 属性の設定に使うハッシュ
+  # @return [Hash] 指定したハッシュ
+  def set_attributes_with_result_page_params(params)
+    self.channels = params['channels']&.split(' ') || []
+    self.since = params['since']
+    self.until = params['until']
+  end
+
+  # 検索結果を返す
+  # @return [MessagePeriodResult] 検索結果
+  # @return [nil] 属性が正しくなかった場合
+  def result
+    return nil unless valid?
+
+    channels = @channels.empty? ?
+      [] : Channel.where(identifier: @channels.split).to_a
+
+    messages = Message.
+      filter_by_channels(channels).
+      filter_by_since(@since).
+      filter_by_until(@until).
+      order(timestamp: :asc, id: :asc).
+      limit(ResultLimit + 1).
+      includes(:channel, :irc_user).
+      to_a
+
+    @until = messages.last.timestamp if messages.count > ResultLimit
+
+    conversation_messages = ConversationMessage.
+      filter_by_channels(channels).
+      filter_by_since(@since).
+      filter_by_until(@until).
+      order(timestamp: :asc, id: :asc).
+      limit(ResultLimit).
+      includes(:channel, :irc_user)
+
+    i = 0
+    result_messages =
+      (messages.to_a + conversation_messages.to_a).
+      sort_by { |m| [m.timestamp, i += 1] }.
+      first(ResultLimit)
+
+    # ソートしたメッセージから、ConversationMessage だけ抽出する
+    result_conversation_messages = result_messages.select do |m|
+      m.class.superclass == ConversationMessage
+    end
```

Comment:
`Enumerable#grep` で同じ動作にできたと思います（（1）`Enumerable#grep` は `===` でマッチするかを判定、（2）`klass === object` で `object.is_a?(klass)` と同様）。

https://docs.ruby-lang.org/ja/3.0.0/class/Enumerable.html#I_GREP

```suggestion
    result_conversation_messages = result_messages.grep(ConversationMessage)
```

Reply:
こういう書き方が出来たのですね、採用させて頂きます。